### PR TITLE
HM-180 Update Api khoa

### PR DESCRIPTION
**Pull request by huynhlap13**

### CODE REVIEW

File: app/Services/KhoaService.php [hunk 1]
```diff
@@ -32,4 +32,40 @@ public function getTreeListKhoaPhong($limit, $page, $benhVienId)
...
+    {
+        $this->khoaRepository->update($id, $input);
+    }
+    
+    public function delete($id)
+    {
+        $this->khoaRepository->delete($id);
+    }
+    
+    public function searchKhoaByKeywords($keyWords)
```
Comment (thien-dang-robustaeng) [inline app/Services/KhoaService.php]: searchKhoaByKeywords->searchByKeywords

File: app/Repositories/KhoaRepository.php [hunk 2]
```diff
@@ -91,4 +91,77 @@ public function getTreeListKhoaPhong($limit = 100, $page = 1, $benhVienId)
...
+    
+    public function delete($id)
+    {
+        $result = $this->find($id);
+        if ($result) {
+            $result->destroy($id);
+        }
+    }
+    
+    public function searchKhoaByKeywords($keyWords)
```
Comment (thien-dang-robustaeng) [inline app/Repositories/KhoaRepository.php]: searchKhoaByKeywords->searchByKeywords

File: app/Repositories/KhoaRepository.php [hunk 3]
```diff
@@ -91,4 +91,77 @@ public function getTreeListKhoaPhong($limit = 100, $page = 1, $benhVienId)
...
+    {
+        $data = $this->model
+                        ->whereRaw('LOWER(ten_khoa) LIKE ? ',['%'.strtolower($keyWords).'%'])
+                        ->get();
+        return $data;
+    } 
+    
+    public function getAllKhoaByBenhVienId($benhVienId)
```
Comment (thien-dang-robustaeng) [inline app/Repositories/KhoaRepository.php]: getAllKhoaByBenhVienId->getAllByBenhVienId

File: app/Services/KhoaService.php [hunk 4]
```diff
@@ -32,4 +32,40 @@ public function getTreeListKhoaPhong($limit, $page, $benhVienId)
...
+    {
+        $data = $this->khoaRepository->searchKhoaByKeywords($keyWords);
+        return $data;
+    }  
+    
+    public function getAllKhoaByBenhVienId($benhVienId)
```
Comment (thien-dang-robustaeng) [inline app/Services/KhoaService.php]: getAllKhoaByBenhVienId->getAllByBenhVienId

File: routes/api.php [hunk 5]
```diff
@@ -299,6 +299,15 @@
         Route::get('getListPhieuChamSocByDieuTriId/{dieuTriId}','PhieuChamSoc\PhieuChamSocController@getAllByDieuTriId');
     });    
     
+    Route::group(['prefix' => 'khoa'], function () {
+		Route::get('getPartial','KhoaPhong\KhoaController@getPartial');
+		Route::post('create','KhoaPhong\KhoaController@create');
+     	Route::post('update/{id}','KhoaPhong\KhoaController@update');
+     	Route::delete('delete/{id}','KhoaPhong\KhoaController@delete');
+     	Route::get('getAllKhoaByBenhVienId/{benhVienId}','KhoaPhong\KhoaController@getAllKhoaByBenhVienId');
```
Comment (thien-dang-robustaeng) [inline routes/api.php]: Route::get('getAllByBenhVienId/{benhVienId}','KhoaPhong\KhoaController@getAllByBenhVienId');

File: routes/api.php [hunk 6]
```diff
@@ -299,6 +299,15 @@
...
+ 		Route::get('searchKhoaByKeywords/{keyWords}','KhoaPhong\KhoaController@searchKhoaByKeywords');
```
Comment (thien-dang-robustaeng) [inline routes/api.php]: Route::get('searchByKeywords/{keyWords}','KhoaPhong\KhoaController@searchByKeywords');

File: app/Services/KhoaService.php [hunk 7]
```diff
@@ -5,13 +5,17 @@
...
 use App\Repositories\KhoaRepository;
+
+use App\Repositories\DanhMuc\DanhMucTongHopRepository;
+
 use Illuminate\Http\Request;
 use Validator;
 
 class KhoaService {
-    public function __construct(KhoaRepository $khoaRepository)
+    public function __construct(KhoaRepository $khoaRepository,DanhMucTongHopRepository $danhmucTongHopRepository)
```
Comment (kptpro) [inline app/Services/KhoaService.php]: $danhmucTongHopRepository -> $danhMucTongHopRepository

File: app/Services/KhoaService.php [hunk 8]
```diff
@@ -5,13 +5,17 @@
...
     {
         $this->khoaRepository = $khoaRepository;
+        $this->danhmucTongHopRepository = $danhmucTongHopRepository;
```
Comment (kptpro) [inline app/Services/KhoaService.php]: $this->danhMucTongHopRepository = $danhMucTongHopRepository

File: app/Services/KhoaService.php [hunk 9]
```diff
@@ -68,4 +72,9 @@ public function getAllByBenhVienId($benhVienId)
         return $data;
     }  
     
+    public function getAllByLoaiKhoa($khoa)
+    {
+        $data = $this->danhmucTongHopRepository->getAllByKhoa($khoa);
```
Comment (kptpro) [inline app/Services/KhoaService.php]: $this->danhmucTongHopRepository sửa lại thành $this->danhMucTongHopRepository

File: app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        return $this->respond($data);
+    }
+    
+    public function getAllByLoaiKhoa($khoa)
+    {
+        $data = $this->khoaService->getAllByLoaiKhoa($khoa);
+        return $this->respond($data);
+    }
+    
+    public function getKhoaById($id)
```
Comment (kptpro) [inline app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php]: getKhoaById -> getById

File: app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        $isNumericId = is_numeric($id);
+        
+        if($isNumericId) {
+            $data = $this->khoaService->getKhoaById($id);
```
Comment (kptpro) [inline app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php]: getKhoaById -> getById

File: app/Http/Controllers/Api/V1/KhoaPhong/PhongController.php [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        return $this->respond($data);
+    }
+    
+    public function getAllByKhoaId($khoaId)
+    {
+        $data = $this->phongService->getAllByKhoaId($khoaId);
+        return $this->respond($data);
+    }
+   
+    public function getPhongById($id)
```
Comment (kptpro) [inline app/Http/Controllers/Api/V1/KhoaPhong/PhongController.php]: getPhongById -> getById

File: app/Http/Controllers/Api/V1/KhoaPhong/PhongController.php [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        $isNumericId = is_numeric($id);
+        
+        if($isNumericId) {
+            $data = $this->phongService->getPhongById($id);
```
Comment (kptpro) [inline app/Http/Controllers/Api/V1/KhoaPhong/PhongController.php]: getPhongById -> getById

File: app/Http/Controllers/Api/V1/KhoaPhong/PhongController.php [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        } else {
+            $this->setStatusCode(400);
+            $data = [];
+        }
+        
+        return $this->respond($data);
+    }
+    
+    public function getAllByLoaiPhong($khoa)
```
Comment (kptpro) [inline app/Http/Controllers/Api/V1/KhoaPhong/PhongController.php]: getAllByLoaiPhong($khoa) -> getAllByLoaiPhong($loaiPhong)

File: app/Http/Controllers/Api/V1/KhoaPhong/PhongController.php [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        $data = $this->phongService->getAllByLoaiPhong($khoa);
```
Comment (kptpro) [inline app/Http/Controllers/Api/V1/KhoaPhong/PhongController.php]: getAllByLoaiPhong($khoa) -> getAllByLoaiPhong($loaiPhong)

File: app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+        return $this->respond($data);
+    }
+    
+    public function getAllByBenhVienId($benhVienid)
+    {
+        $data = $this->khoaService->getAllByBenhVienId($benhVienid);
+        return $this->respond($data);
+    }
+    
+    public function getAllByLoaiKhoa($khoa)
```
Comment (kptpro) [inline app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php]: getAllByLoaiKhoa($khoa) -> getAllByLoaiKhoa($loaiKhoa)

File: app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        $data = $this->khoaService->getAllByLoaiKhoa($khoa);
```
Comment (kptpro) [inline app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php]: getAllByLoaiKhoa($khoa) -> getAllByLoaiKhoa($loaiKhoa)

File: app/Http/Requests/KhoaFormRequest.php [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'ma_khoa'                   => 'required|string',
+            'ten_khoa'                   => 'required|string',
+            'loai_khoa'                => 'nullable|int',
```
Comment (kptpro) [inline app/Http/Requests/KhoaFormRequest.php]: loai_khoa is required

File: app/Http/Requests/PhongFormRequest.php [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'khoa_id'              => 'required|int',
+            'so_phong'              => 'required|int',
+            'ma_nhom'                   => 'required|string',
+            'ten_phong'                   => 'required|string',
+            'loai_phong'                => 'nullable|int'
```
Comment (kptpro) [inline app/Http/Requests/PhongFormRequest.php]: loai_phong is required

File: app/Repositories/KhoaRepository.php [hunk 20]
```diff
@@ -91,4 +91,120 @@ public function getTreeListKhoaPhong($limit = 100, $page = 1, $benhVienId)
...
+    
+    public function getAllByBenhVienId($benhVienId)
+    {
+        $data = $this->model
+                    ->where('benh_vien_id', $benhVienId)
+                    ->get();
+        return $data;
+    } 
+    
+    public function getKhoaById($id)
```
Comment (kptpro) [inline app/Repositories/KhoaRepository.php]: getKhoaById -> getById

File: app/Repositories/PhongRepository.php [hunk 21]
```diff
@@ -73,4 +73,113 @@ public function getPhongNoiTruByKhoaId($khoaId) {
...
+    
+    public function getAllByKhoaId($khoaId)
+    {
+        $data = $this->model
+                    ->where('khoa_id', $khoaId)
+                    ->get();
+        return $data;
+    } 
+    
+    public function getPhongById($id)
```
Comment (kptpro) [inline app/Repositories/PhongRepository.php]: getPhongById -> getById

File: app/Services/KhoaService.php [hunk 22]
```diff
@@ -31,4 +35,52 @@ public function getTreeListKhoaPhong($limit, $page, $benhVienId)
...
+        return $data;
+    }  
+    
+    public function getAllByLoaiKhoa($khoa)
+    {
+        $data = $this->danhmucTongHopRepository->getAllByKhoa($khoa);
+        return $data;
+    }  
+    
+    public function getKhoaById($id)
```
Comment (kptpro) [inline app/Services/KhoaService.php]: getKhoaById -> getById

File: app/Services/KhoaService.php [hunk 23]
```diff
@@ -31,4 +35,52 @@ public function getTreeListKhoaPhong($limit, $page, $benhVienId)
...
+    {
+        $data = $this->khoaRepository->getKhoaById($id);
```
Comment (kptpro) [inline app/Services/KhoaService.php]: getKhoaById -> getById

File: app/Services/PhongService.php [hunk 24]
```diff
@@ -25,4 +30,58 @@ public function getNhomPhong($loaiPhong,$khoaId)
...
+        return $data;
+    }  
+    
+    public function getAllByKhoaId($khoaId)
+    {
+        $data = $this->phongRepository->getAllByKhoaId($khoaId);
+        return $data;
+    }  
+    
+    public function getPhongById($id)
```
Comment (kptpro) [inline app/Services/PhongService.php]: getPhongById -> getById

File: app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+        return $this->respond([]);        
+    }
+    
+    public function searchByKeywords($keyWords)
+    {
+        $data = $this->khoaService->searchByKeywords($keyWords);
+        return $this->respond($data);
+    }
+    
+    public function getAllByBenhVienId($benhVienid)
```
Comment (kptpro) [inline app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php]: $benhVienid -> $benhVienId

File: app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        $data = $this->khoaService->getAllByBenhVienId($benhVienid);
```
Comment (kptpro) [inline app/Http/Controllers/Api/V1/KhoaPhong/KhoaController.php]: $benhVienid -> $benhVienId

File: app/Services/PhongService.php [hunk 27]
```diff
@@ -5,13 +5,18 @@
...
+
 use Illuminate\Http\Request;
 use Validator;
 
 class PhongService {
-    public function __construct(PhongRepository $phongRepository)
+    public function __construct(PhongRepository $phongRepository,DanhMucTongHopRepository $danhmucTongHopRepository,KhoaRepository $khoaRepository)
     {
         $this->phongRepository = $phongRepository;
+        $this->danhmucTongHopRepository = $danhmucTongHopRepository;
```
Comment (kptpro) [inline app/Services/PhongService.php]: $this->danhMucTongHopRepository = $danhMucTongHopRepository;

File: app/Services/PhongService.php [hunk 28]
```diff
@@ -5,13 +5,18 @@
...
 use App\Repositories\PhongRepository;
+use App\Repositories\DanhMuc\DanhMucTongHopRepository;
+use App\Repositories\KhoaRepository;
+
 use Illuminate\Http\Request;
 use Validator;
 
 class PhongService {
-    public function __construct(PhongRepository $phongRepository)
+    public function __construct(PhongRepository $phongRepository,DanhMucTongHopRepository $danhmucTongHopRepository,KhoaRepository $khoaRepository)
```
Comment (kptpro) [inline app/Services/PhongService.php]: $danhmucTongHopRepository -> $danhMucTongHopRepository

**Pull request merged**